### PR TITLE
Bill Users for Paid Accounts

### PR DESCRIPTION
**Pull request by stephenprater**
## What changes does this PR introduce?

This enables us to bill users through Stripe for Paid Accounts

## Any background context you want to provide?

There are a few major parts to this PR:
1) Upload Plan data to stripe consistently
2) Add "foreign keys" to models to maintain their connection to stripe.
3) Allow users to update their records via the API
4) Keep those changes in sync with Stripe
5) Respond to Stripe initiated changes via webhook and keep those changes in sync with our tables.

As an outgrowth of 3/4/5 I discovered that our use of SQLAlchemy made the models both hard to test in isolation and likely to result in a lot of specious calls to stripe.  We needed better controls of the change management than SQLAlchemy's (coarse) callbacks offered, so we need to track changes ourselves.  About half of our interaction methods need to either flush or commit to the DB, but it's it's confusing to have a lot of database state management mixed in with business logic.

To avoid that, I created an `Interactor` mixin that let's us manage that kind of housekeeping a little more decoratively.

## Where should the reviewer start?

I tried to order these commits in a logical order so it made sense when you read them instead of a single 5,000 line diff - so you should be able to start at the top.

## Has this been manually tested? How?

Tests pass!  This one is _definitely_ going to need pushbutton.

## What value does this provide to our end users?

Let's them pay us for our awesome service, keeping us all employed and our service running!

## What GIF best describes this PR or how it makes you feel?
making it rain

### CODE REVIEW

**Comment by stephenprater** [comment]
[Allow Users to Change their Plans](https://trello.com/c/1Xg3Nix5/487-allow-users-to-change-their-plans)

File: app/jobs/email.py [hunk 1]
```diff
@@ -120,19 +130,187 @@ def send_password_reset_confirm_email(email_address, token_url):
...
+    email.body_html = f"""
+<html>
+<head></head>
+<body>
+    <h1>
+        Welcome to Holepunch!
+    </h1>
+    <p>Thanks for subscribing!  Your Holepunch.io Account is now Active</p>
+    <p>If you didn't make this change, please let us know at support@holepunch.io</p>
+    <p>Thanks!</p>
```
Comment (AndrewScibek) [inline app/jobs/email.py]: If you have any questions about your subscription...

File: app/utils/general.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import collections
+import functools
+
+
+class memoized(object):
```
Comment (AndrewScibek) [inline app/utils/general.py]: This is repeated in json.py

**Comment by stephenprater** [comment]
[Allow Users to Change their Plans](https://trello.com/c/1Xg3Nix5/487-allow-users-to-change-their-plans)

**Pull request merged**